### ISSUE TITLE
About the unstable performance of the class train on seed 2

### ISSUE THREAD

**Issue by KevinChen880723** [lhoyer/MIC#9]
Dear author,

First of all, thanks for your contributions to the deep learning community :)
After reproducing the experiments, I was able to achieve the same results as recorded in the paper with some random seeds, but sometimes wasn't: 
- 76.0 (seed=1)
- 72.47 (seed=2)
- 75.86 (seed=3)

I find the bad performance with random seed=2 is because of the low IoU on "train", which is near 0 at 40k iterations and only 30.48 at 60k iterations.
Have you faced this issue before? 
If so, what do you think might be the reason for it?

Thank you!
YuKai Chen

**Comment by lhoyer** [comment]
Dear YuKai,

Thank you for your interest in our work! On our cluster, all tested seeds (0, 1, and 2) yielded a stable performance around 75.9 mIoU. It seems that you have trained the experiments with 60k iterations. This could affect the stability as it affects the learning rate schedule, so that the model is trained longer with a higher learning rate, which might increase instabilities. Have you tried to train seed 2 with 40k max_iters?

Apart from this possible reason, @zyuanbing has observed a similar instability for 40k iterations and seed 2 in https://github.com/lhoyer/MIC/issues/8#issuecomment-1418435917. On our cluster, we did not observe this instability with seed 2. My current hypothesis is that due to `torch.cudnn.benchmark = True`, another set of implementation is selected on other machines (see https://pytorch.org/docs/1.7.1/notes/randomness.html#cuda-convolution-benchmarking). As the decision boundary for rare classes such as `train` can be quite brittle (`train` can be easily confused with `bus`), certain implementations could favor these instabilities. Unfortunately, it is not possible to copy the used implementations from one machine to another. This issue might be more severe for MIC+HRDA than for HRDA as MIC might mask out `train` in unfavorable situations. This would also explain why only seed 2 seems to be affected. A possible mitigation strategy could be reducing the mask ratio for rare classes (based on the pseudo-labels). 

Best,
Lukas

**Comment by KevinChen880723** [comment]
Dear Lukas,

Thank you for your response!

So to conclude, could we say that a good random mask is crucial for MIC's performance?
Although the performance of MIC is generally excellent, its sensitivity to randomness can sometimes cause instability. 
To address this issue, reducing the mask ratio may be a potential solution.

Thank you for sharing your valuable perspective :)

Best regards,
YuKai Chen
